sycl/plugins/unified_runtime/ur/adapters/level_zero/command_buffer.cpp

```diff
@@ -814,6 +815,22 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferEnqueueExp(
     ZE2UR_CALL(zeCommandListAppendBarrier,
                (SignalCommandList->first, RetEvent->ZeEvent, 1,
                 &(CommandBuffer->SignalEvent->ZeEvent)));
+
+    // We create an extra-signal specific to the current execution of the
+    // CommandBuffer. This signal is needed for profiling the execution time
+    // of the CommandBuffer. It waits for the WaitEvent to be signaled
+    // which indicates the start of the CommandBuffer actual execution.
+    // This event is embeded into the Event return to user to allow
+    // the profiling engine to retrieve it.
+    UR_CALL(createEventAndAssociateQueue(Queue, &StartEvent,
+                                         UR_COMMAND_COMMAND_BUFFER_ENQUEUE_EXP,
+                                         WaitCommandList, false));
+
+    ZE2UR_CALL(zeCommandListAppendBarrier,
+               (WaitCommandList->first, StartEvent->ZeEvent, 1,
+                &(CommandBuffer->WaitEvent->ZeEvent)));
+
+    RetEvent->CommandData = StartEvent;
```

Comment:
Do avoid doing this work all the time, can we only do it if the queue has `UR_QUEUE_FLAG_PROFILING_ENABLE` set?

Might also be worth adding something to the design doc about this approach https://github.com/reble/llvm/blob/sycl-graph-develop/sycl/doc/design/CommandGraph.md#level-zero (will be a faff to update the diagram so can just put in some text)

Reply:
I added the condition and updated the design doc